### PR TITLE
File cache umask

### PR DESCRIPTION
**Pull request by bshaffer**
see #806

### CODE REVIEW

**Comment by bshaffer** [comment]
@huebs please comment here to make the CLA bot happy :D

**Comment by bshaffer** [comment]
@tmatsuo PTAL

File: src/Google/Cache/File.php [hunk 1]
```diff
@@ -155,7 +155,7 @@ private function getCacheDir($file, $forWrite)
     // trim the directory separator from the path to prevent double separators
     $storageDir = rtrim($this->path, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR . $dirHash;
     if ($forWrite && ! is_dir($storageDir)) {
-      if (! mkdir($storageDir, 0700, true)) {
+      if (!mkdir($storageDir, 0777, true)) {
```
Comment (tmatsuo) [inline src/Google/Cache/File.php]: 0777 seems little bit dangerous. What kind of data will be in the cache? Is any sensitive data there? Does it make sense to use 0770 instead?

File: src/Google/Cache/File.php [hunk 2]
```diff
@@ -199,7 +199,7 @@ private function acquireLock($type, $storageFile)
       return false;
     }
     if ($type == LOCK_EX) {
-      chmod($storageFile, 0600);
+      chmod($storageFile, 0666 & ~umask());
```
Comment (tmatsuo) [inline src/Google/Cache/File.php]: `0660` here too

File: src/Google/Cache/File.php (hunk 1 above)
Comment (tmatsuo) [inline src/Google/Cache/File.php]: Ah nm, it will be umasked anyways, so I think it's fine.

File: src/Google/Cache/File.php (hunk 2 above)
Comment (tmatsuo) [inline src/Google/Cache/File.php]: nm too

**Comment by tmatsuo** [comment]
LGTM

File: src/Google/Cache/File.php (hunk 1 above)
Comment (bshaffer) [inline src/Google/Cache/File.php]: Access tokens are stored in cache, yes. But the whole point is to fix #806. So we either break some of our users and be extra secure, or choose to be standard with other file caches and make file permissions looser.

Comment (tmatsuo) [inline src/Google/Cache/File.php]: Ok, let me think about it.

Comment (tmatsuo) [inline src/Google/Cache/File.php]: I think the root user's cache should not be accessible from other users.
Does it make sense to prefix the cache directory with the username?

**Comment by huebs** [comment]
Making bots happy. :smiley:

**Comment by tmatsuo** [comment]
Oh, I LGTMed, but did undo LGTM. Can you revert? See https://github.com/google/google-api-php-client/pull/839#discussion_r52365593

File: src/Google/Cache/File.php (hunk 1 above)
Comment (huebs) [inline src/Google/Cache/File.php]: @tmatsuo, as I mentioned in #817:

> I foresee two problems with using hashed usernames as directory names for namespacing.
>
> - CLI cache cleaners will be unable to remove files created by the web interface.
> - The web interface will not be able to read files created by a CLI cache warmer.

Comment (tmatsuo) [inline src/Google/Cache/File.php]: > CLI cache cleaners will be unable to remove files created by the web interface.

Why do you need to cleanup?

> The web interface will not be able to read files created by a CLI cache warmer.

Is this a big problem?

Think about this scenario.

I have downloaded the credentials file and stored secretly, only available for root.
root process will create a cache which contains the access token from the credentials.

Then any other users should not be able to read that cache.

The correct approach for the cron scenario is to run the process as an appropriate user, not root.

File: tests/Google/CacheTest.php [hunk 3]
```diff
@@ -40,6 +40,48 @@ public function testFileWithTrailingSlash()
...
+
+    $method = new ReflectionMethod($cache, 'getWriteableCacheFile');
+    $method->setAccessible(true);
+    $filename = $method->invoke($cache, 'foo');
+    $stat = stat($filename);
+
+    $this->assertEquals(0666 & ~umask(), $stat['mode'] & 0777);
+  }
+
+  public function testFileWithCustomMask()
```
Comment (huebs) [inline tests/Google/CacheTest.php]: This test case can be removed as it now just duplicates the previous one.

File: src/Google/Cache/File.php (hunk 1 above)
Comment (bshaffer) [inline src/Google/Cache/File.php]: Added username hash. see #840

Comment (bshaffer) [inline src/Google/Cache/File.php]: Agreed with @tmatsuo this is not a general-purpose file cache. it is a credentials cache. So we'll stay on the safe side here.

**Pull request merged**